server/src-lib/Hasura/GraphQL/Schema.hs

```diff
@@ -1,960 +1,598 @@
+        pure $ fmap (RFDB . MDBUpdate) <$> catMaybes [update, updateByPk]
+
+      deletes <- fmap join $ whenMaybe (isMutable viIsDeletable viewInfo) $ for (_permDel permissions) \deletePerms -> do
+        let deleteName = $$(G.litName "delete_") <> displayName
+            deleteDesc = G.Description $ "delete data from the table: " <>> table
+            deleteByPkName = $$(G.litName "delete_") <> displayName <> $$(G.litName "_by_pk")
+            deleteByPkDesc = G.Description $ "delete single row from the table: " <>> table
+        delete <- deleteFromTable table (fromMaybe deleteName $ _tcrfDelete customRootFields) (Just deleteDesc) deletePerms selectPerms
+
+        -- ditto
+        deleteByPk <- join <$> for selectPerms
+          (deleteFromTableByPk table (fromMaybe deleteByPkName $ _tcrfDeleteByPk customRootFields) (Just deleteByPkDesc) deletePerms)
+        pure $ fmap (RFDB . MDBDelete) delete : maybe [] (pure . fmap (RFDB . MDBDelete)) deleteByPk
+
+      pure $ concat $ catMaybes [inserts, updates, deletes]
+
+  actionParsers <- for allActions $ \actionInfo ->
+    case _adType (_aiDefinition actionInfo) of
+      ActionMutation ActionSynchronous ->
+        fmap (fmap (RFAction . AMSync)) <$> actionExecute nonObjectCustomTypes actionInfo
+      ActionMutation ActionAsynchronous ->
+        fmap (fmap (RFAction . AMAsync)) <$> actionAsyncMutation nonObjectCustomTypes actionInfo
+      ActionQuery -> pure Nothing
+
+  let mutationFieldsParser = concat (catMaybes mutationParsers) <> catMaybes actionParsers <> fmap (fmap RFRemote) allRemotes
+  pure if null mutationFieldsParser
+       then Nothing
+       else Just $ P.selectionSet $$(G.litName "mutation_root") (Just $ G.Description "mutation root") mutationFieldsParser
+            <&> fmap (P.handleTypename (RFRaw . J.String . G.unName))
+
+unauthenticatedMutation
+  :: forall n
+   . MonadParse n
+  => [P.FieldParser n (RemoteSchemaInfo, G.Field G.NoFragments P.Variable)]
+  -> Maybe (Parser 'Output n (OMap.InsOrdHashMap G.Name (MutationRootField UnpreparedValue)))
+unauthenticatedMutation allRemotes =
+  let mutationFieldsParser = fmap (fmap RFRemote) allRemotes
+  in if null mutationFieldsParser
+     then Nothing
+     else Just $ P.selectionSet $$(G.litName "mutation_root") Nothing mutationFieldsParser
+          <&> fmap (P.handleTypename (RFRaw . J.String . G.unName))
+
+unauthenticatedSubscription
+  :: forall n
+   . MonadParse n
+  => Parser 'Output n (OMap.InsOrdHashMap G.Name (QueryRootField UnpreparedValue))
+unauthenticatedSubscription =
+ P.selectionSet $$(G.litName "subscription_root") Nothing []
+ <&> fmap (P.handleTypename (RFRaw . J.String . G.unName))
+
+checkFieldNamesUnique
+  :: forall m
+   . ( Monad m
+     )
+  => [G.Name]
+  -- ^ list of fields whose names are to be checked for uniqueness
+  -> (G.Name -> m ())
+  -- ^ error action
+  -> m ()
+checkFieldNamesUnique fields err = do
```

Comment:
Seems overkill, can probably be replaced by a call to `Data.List.Extended.duplicates`.

Reply:
Agreed, I was being sloppy here.

Reply:
Got rid of `checkFieldNamesUnique`, replaced with calls to `duplicates`. Resolving.